lib/TextDocument/TextDocumentUri.php

```diff
@@ -39,35 +42,46 @@ public static function fromString(?string $uri): self
 
         $match = preg_match('{^(?<scheme>[a-z]+://){0,1}(?<path>.+)?}', $uri, $components);
 
-        if (!isset($components['scheme']) || $components['scheme'] == '') {
-            $components['scheme'] = self::SCHEME_FILE;
-        } else {
-            $components['scheme'] = substr($components['scheme'], 0, -3);
-        }
-
         if (!isset($components['path'])) {
             throw new InvalidUriException(sprintf(
                 'URI "%s" has no path component',
                 $uri
             ));
         }
 
-        if (!in_array($components['scheme'], self::SCHEMES)) {
-            throw new InvalidUriException(sprintf(
-                'Only "%s" schemes are supported, got "%s"',
-                implode('", "', self::SCHEMES),
-                $components['scheme']
-            ));
-        }
+        if (!isset($components['scheme']) || $components['scheme'] == '') {
+            // Allow this function to accept filesystem paths too (not URIs), convert to file: URIs
+            $components['scheme'] = self::SCHEME_FILE;
 
-        if ($components['scheme'] === self::SCHEME_FILE && false === Path::isAbsolute($uri)) {
-            throw new InvalidUriException(sprintf(
-                'URI for file:// must be absolute, got "%s"',
-                $uri
-            ));
-        }
+            if ($components['scheme'] === self::SCHEME_FILE && !Path::isAbsolute($uri)) {
+                throw new InvalidUriException(sprintf(
+                    'Path for file:// must be absolute, got "%s"',
+                    $uri
+                ));
+            }
 
-        $components['path'] = Path::canonicalize($components['path']);
+            $components['path'] = Path::canonicalize($components['path']);
+
+        } else {
```

Comment:
note in general I avoid using `else` in this code base in favour of _early returns_ - I don't want to make you jump through hoops, but do you think you could refactor to remove the `else`? otherwise happy to merge it and we can avoid it in the future.

Reply:
Sure, I don't really agree with that blog post, but I'm happy to follow your conventions. I'll try to keep it in mind for any future patches.